src/os/linux/mod.rs

```diff
@@ -0,0 +1,165 @@
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum OpenFileError {
+    LaunchFailed(bossy::Error),
+}
+
+impl Display for OpenFileError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::LaunchFailed(e) => write!(f, "Launch failed: {}", e),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Application {
+    exec_command: OsString,
+    icon: Option<OsString>,
+    xdg_entry_path: PathBuf,
+}
+
+impl Application {
+    pub fn detect_editor() -> Result<Self, DetectEditorError> {
+        let entry = {
+            // Try a rust code editor, then a plain text editor, if neither are available,
+            // then return an error.
+            if let Some(rust_entry) = xdg::query_mime_entry("text/rust") {
+                rust_entry
+            } else if let Some(plain_entry) = xdg::query_mime_entry("text/plain") {
+                plain_entry
+            } else {
+                return Err(DetectEditorError::NoDefaultEditorSet);
+            }
+        };
+
+        for dir in xdg::get_xdg_data_dirs() {
+            // Look at the applications folder of that dir
+            let dir = dir.join("applications");
+            if let Some(result) = xdg::find_entry_in_dir(&dir, &entry) {
+                let parsed = xdg::parse(&result)
+                    .ok_or(DetectEditorError::FreeDesktopEntryParseError)?;
+                return Ok(Self {
+                    exec_command: parsed
+                        .section("Desktop Entry")
+                        .attr("Exec")
+                        .ok_or(DetectEditorError::FreeDesktopEntryParseError)?
+                        .into(),
+                    icon: parsed
+                        .section("Desktop Entry")
+                        .attr("Icon")
+                        .map(|s| s.into()),
+                    xdg_entry_path: dir.join(result),
+                })
+            }
+        }
+
+        Err(DetectEditorError::FreeDesktopEntryNotFound)
+    }
+
+    pub fn open_file(&self, path: impl AsRef<Path>) -> Result<(), OpenFileError> {
+        let path = path.as_ref();
+
+        let maybe_icon = if let Some(icon_str) = &self.icon {
+            Some(icon_str.as_os_str())
+        } else {
+            None
+        };
+
+        // Parse the xdg command field with all the needed data
+        let command_parts = xdg::parse_command(
+            &self.exec_command,
+            path.as_os_str(),
+            maybe_icon,
+            Some(&self.xdg_entry_path)
+        );
+
+        if !command_parts.is_empty() {
+            // If command_parts has at least one element this works. If it has a single
+            // element, &command_parts[1..] should be an empty slice (&[]) and bossy
+            // `add_args` does not add any argument on that case, although the docs
+            // do not make it obvious.
+            bossy::Command::impure(&command_parts[0])
+                .add_args(&command_parts[1..])
```

Comment:
```suggestion
                .with_args(&command_parts[1..])
```
The `add_*` methods are only necessary if we need to mutate by reference.